lib/cn/route.c

```diff
@@ -0,0 +1,247 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#include <hse_ikvdb/kvs_cparams.h>
+
+#include <hse_util/platform.h>
+#include <hse_util/alloc.h>
+#include <hse_util/logging.h>
+#include <hse_util/assert.h>
+#include <hse_util/keycmp.h>
+#include <hse_util/byteorder.h>
+
+#include <rbtree.h>
+
+#include "route.h"
+
+struct route_node {
+    union {
+        struct rb_node rtn_node;
+        struct route_node *rtn_next;
+    };
+    uint16_t       rtn_keylen;
+    uint16_t       rtn_child;
+    uint8_t        rtn_keybuf[28];
+};
+
+struct route_map {
+    struct rb_root     rtm_root;
+    uint               rtm_pfxlen;
+    uint               rtm_fanout;
+    struct route_node *rtm_free;
+    struct route_node  rtm_nodev[] HSE_L1D_ALIGNED;
+};
+
+static struct route_node *
+route_map_insert(struct route_map *map, struct route_node *node)
+{
+    struct rb_root *root = &map->rtm_root;
+    struct rb_node **link = &root->rb_node;
+    struct rb_node *parent = NULL;
+
+    while (*link) {
+        struct route_node *this;
+        int rc;
+
+        this = rb_entry(*link, struct route_node, rtn_node);
+        parent = *link;
+
+        rc = keycmp(node->rtn_keybuf, node->rtn_keylen, this->rtn_keybuf, this->rtn_keylen);
+
+        if (rc < 0) {
+            link = &(*link)->rb_left;
+        } else if (rc > 0) {
+            link = &(*link)->rb_right;
+        } else {
+            return this;
+        }
+    }
+
+    rb_link_node(&node->rtn_node, parent, link);
+    rb_insert_color(&node->rtn_node, root);
+
+    return NULL;
+}
+
+static struct route_node *
+route_map_find(struct route_map *map, const void *key, uint keylen)
+{
+    struct rb_root *root = &map->rtm_root;
+    struct rb_node **link = &root->rb_node;
+    struct rb_node *node = *link;
+    uint len = keylen;
+    int dir = 0;
+
+    if (len > map->rtm_pfxlen)
+        len = map->rtm_pfxlen;
+
+    while (*link) {
+        struct route_node *this;
+        int rc;
+
+        this = rb_entry(*link, struct route_node, rtn_node);
+
+        rc = memcmp(key, this->rtn_keybuf, len);
+
+        if (rc < 0) {
+            dir = -1;
```

Comment:
"dir" is used to keep track of the inorder successor of the searched key, right?

Reply:
Yes, I need to keep track of whether I've ever gone left in order to choose the correct node for cases where we don't have an exact edge key match and hence fall out the bottom of the loop.
Was maddening trying to figure this out.  Someone should review it carefully!!!

Reply:
This makes sense to me..
The last node from which a left turn was taken in the traversal path is the inorder successor of the searched key.
If there was no left turn taken in the traversal path, then the inorder successor is the last parent node.